src/PhotonBackground.cpp

```diff
@@ -63,64 +66,377 @@ static PhotonFieldScaling scalingStecker16_upper("IRB_Stecker16_upper");
+        } else {
+            ss2 = 26.4*std::pow(s, -0.34) + 59.3*std::pow(s, 0.095);
+        }
+        cs_multidiff = (1.-std::exp(-ss1))*ss2;
+        cs_multi = 0.89*cs_multidiff;
+        // diffractive scattering:
+        cross_diffr1 = .099*cs_multidiff;
+        cross_diffr2 = .011*cs_multidiff;
+        cross_diffr = .11*cs_multidiff;
+        // **************************************
+        ss1 = std::pow((x-.85), .75)/.64;
+        ss2 = 74.1*std::pow(x, -0.44) + 62.*std::pow(s, 0.08);
+        double cs_tmp = 0.96*(1.-std::exp(-ss1))*ss2;
+        cross_diffr1 = 0.14*cs_tmp;
+        cross_diffr2 = 0.013*cs_tmp;
+        double cs_delta = cross_frag2
+                        - (cross_diffr1+cross_diffr2-cross_diffr);
+        if (cs_delta < 0.) {
+            cross_frag2 = 0.;
+            cs_multi += cs_delta;
+        } else {
+            cross_frag2 = cs_delta;
+        }
+        cross_diffr = cross_diffr1 + cross_diffr2;
+        cs_multidiff = cs_multi + cross_diffr;
+    // *****************************************
+    } else {
+        cross_diffr = 0.;
+        cross_diffr1 = 0.;
+        cross_diffr2 = 0.;
+        cs_multidiff = 0.;
+        cs_multi = 0.;
+    }
+    return cross_res+cross_dir+cs_multidiff+cross_frag2;
+}  // end crossection
+
+
+double Photon_Field::Pl(double x, double xth, double xmax, double alpha) const {
+/*
+    - input: photon energy [eV], threshold [eV], max [eV], unknown [no unit]
+    - output: unknown [no unit]
+    - called by: crossection
+*/
+    if (xth > x)
+        return 0.;
+    double a = alpha*xmax/xth;
+    double prod1 = std::pow((x-xth)/(xmax-xth), (a-alpha));
+    double prod2 = std::pow(x/xmax, -a);
+    return prod1*prod2;
+}  // end Pl
+
+
+double Photon_Field::Ef(double x, double th, double w) const {
+/*
+    - input: photon energy [eV], threshold [eV], unknown [eV]
+    - output: unknown [no unit]
+    - called by: crossection
+*/
+    double wth = w+th;
+    if (x <= th) {
+        return 0.;
+    } else if ( (x > th) && (x < wth) ) {
+        return (x-th)/w;
+    } else if (x >= wth) {
+        return 1.;
+    } else {
+        throw std::runtime_error("error in function Ef");
+    }
+}  // end Ef
+
+
+double Photon_Field::breitwigner(double sigma_0,
+                                 double Gamma,
+                                 double DMM,
+                                 double epsPrime,
+                                 bool onProton) const {
+/*
+    - input: cross section [µbarn], width [GeV], mass [GeV/c^2]
+    - output: Breit-Wigner crossection of a resonance widh width Gamma
+    - called by: crossection
+*/
+    const double mass = onProton? 0.93827 : 0.93947;  // Gev/c^2
+    double s = mass*mass + 2.*mass*epsPrime;
+    double gam2s = Gamma*Gamma*s;
+    return sigma_0 * (s/epsPrime/epsPrime)*gam2s
+                   / ((s-DMM*DMM)*(s-DMM*DMM)+gam2s);
+}  // end breitwigner
+
+
+double Photon_Field::functs(double s, bool onProton) const {
```

Comment:
Please use unit conventions everywhere

Reply:
All SOPHIA methods use GeV as their internal standard unit. As not few parts of the original code are very difficult to read and to interpret, all methods directly taken from the original code are marked `SOPHIA_<methodName>` and are being based on their own unit system. Nevertheless, all public methods relating to SOPHIA do use the CRPropa unit convention now - such that the user is not exposed to more than one unit convention.